aioredis/connection.py

```diff
@@ -552,6 +553,105 @@ async def read_response(self) -> Union[EncodableT, List[EncodableT]]:
         return cast(Union[EncodableT, List[EncodableT]], response)
 
 
+@enum.unique
+class _State(enum.IntEnum):
+    not_connected = enum.auto()
+    connected = enum.auto()
+    error = enum.auto()
+
+
+class RedisProtocol(asyncio.Protocol):
+    """ """
+
+    def __init__(self):
+        self._state = _State.not_connected
+        self._resp_queue = deque()
+
+        self._transport: Optional[asyncio.Transport] = None
+
+        self._parser = hiredis.Reader()
+
+        self._responder_task = None
+        self._exc = None
+        self._conn_waiter = asyncio.get_event_loop().create_future()
```

Comment:
I'd recommend using an `asyncio.Event` for this. With a raw future, calling `wait_connected` and then cancelling it will cancel the future. Although it might not be needed at all - see my comments where `create_connection` is called.